### PR TITLE
The delta gap guard sat on the fleet's median sweep gap, so half of every delta collector's output was a zero that never happened

### LINKED ISSUE #2233: [BUG] Every delta collector's 300s gap guard trips on ordinary sweep jitter — 21-45% of deltas silently written as 0
Reported against `get_perfmon_trend` (`delta_value` zeroed while `value` is fine). Reproduced, and the cause is **not** perfmon-specific and **not** an arithmetic error — it is one shared gap guard, and it affects every delta-based metric in the product including wait analysis.

## The measurement

`get_perfmon_trend('Batch Requests/sec')`, prod-sql-use2-alpha-01, 3h window, 38 intervals. Correlating each stored `delta_value` against the wall-clock gap to the previous collection:

```
ZEROED  n=17   gaps 300.9s – 441.7s
CORRECT n=21   gaps 149.4s – 293.6s
```

A clean split at **300 seconds**, with no overlap. And every non-zero delta equals the true difference of the adjacent `value`s **exactly** (verified all 21). So the delta computation is right; the value is being thrown away afterwards.

Same box, same window, `get_wait_trend('SOS_SCHEDULER_YIELD')`: 6/28 zeroed, gaps 302–357s, present 150–294s — **the same 300s boundary, at the same timestamps**. One cause, not two bugs.

| collector | samples | zeroed | rate |
|---|---|---|---|
| perfmon_stats | 38 | 17 | **45%** |
| wait_stats | 28 | 6 | **21%** |

## Mechanism

Every delta call site passes `maxGapSeconds: 300` — 36+ of them across `PerfmonStatsCollector`, `WaitStatsCollector`, `FileIoStatsCollector`, `QueryStatsCollector`, `ProcedureStatsCollector`, `LatchStatsCollector`, `SpinlockStatsCollector`, `MemoryGrantsCollector`, and both PostgreSQL collectors (`PgWaitStatsCollector`, `PgStatementStatsCollector`).

`CollectorDeltaCalculator.CalculateDeltaWithInterval` discards the delta when the gap exceeds it:

```csharp
if (maxGapSeconds > 0 && collectionTime.HasValue && previous.Timestamp.HasValue
    && (collectionTime.Value - previous.Timestamp.Value).TotalSeconds > maxGapSeconds)
{
    delta = 0; interval = 0;
    return (currentValue, collectionTime);
}
```

**The guard's premise no longer holds.** `PerfmonStatsCollector`'s own comment says 300s is *"5x the 1-min collection interval"*, and `CollectorScheduleDefaults` does say `["perfmon_stats"] = new(1, 30)`. But on this 52-server box the observed inter-collection gap for that 1-minute collector is **149–442s** — so 300s is roughly **1x** the real cadence, not 5x, and ordinary sweep jitter lands on both sides of it. The gaps also shrink toward the present (150–230s recently vs 300–440s earlier), which points at sweep contention varying with load rather than a configured interval.

## Why the zero is the worse half of the bug

A discarded delta is written as `0`, which is indistinguishable from *"the counter did not move"*. On `get_wait_trend` that renders as `wait_time_ms_per_second: 0` for a server that was actively accumulating `SOS_SCHEDULER_YIELD` — a confidently wrong answer, not a gap. Any consumer summing or averaging deltas over a window is under-reporting by the zeroed fraction, silently.

Note that `get_wait_trend` **already divides by an interval** to produce a per-second rate. So the guard is discarding a numerator the consumer was going to normalize anyway.

## Options

1. **Raise `maxGapSeconds`.** Cheapest, but a constant just relocates the boundary — a larger fleet has larger gaps and reproduces this. Whatever value is chosen is wrong at some server count.
2. **Derive it from the collector's configured interval** (say 3x `FrequencyMinutes`). Better, but still assumes actual cadence tracks the schedule, which is exactly the assumption that failed here.
3. **Keep the delta and store the interval it accrued over.** `CalculateDeltaWithInterval` already exists and `QueryStatsCollector` already uses it for worker time. A long gap does not make a delta *wrong*, only non-comparable — and the interval is precisely what restores comparability. Consumers then compute rates, which several already do.
4. **At minimum, stop writing `0`.** Distinguish "not computable" from "measured zero", the way `pg_stat_io`'s nullable counters and `write_counters_tracked` distinguish "not measured" from "no writes".

I'd argue (3) + (4): the guard exists to stop an unknown-span delta being read as one interval's worth, and storing the span solves that directly instead of destroying the measurement.

## The measurement that should decide

Whether actual per-collector cadence tracks the configured schedule at fleet scale. If it does not — and the data above says it does not on 52 servers — then any constant or schedule-derived gap threshold is guessing, and (3) is the only option that stays correct as the fleet grows.

Also worth separating out: **why a 1-minute collector is running every ~5 minutes on this box.** That may be entirely reasonable sweep behaviour at 52 servers, but it means the schedule is aspirational, and several comments in the collectors reason from it as though it were real.

Reported by another agent via MCP; the "value column is fine, delta_value is zeroed" framing was accurate and is what made this findable.

### LINKED ISSUE #2234: get_perfmon_trend: delta_value returns a fabricated 0 on ~38% of points while cntr_value advances
## Summary

`get_perfmon_trend` returns `delta_value = 0` on a large fraction of points while the cumulative `value` on those same points advances normally. Any rate derived from `delta_value` is wrong, and — because a spurious 0 is indistinguishable from a real idle interval — it is wrong in a way that reads as a finding.

## Observed

`prod-sql-use2-alpha-01`, counter `Batch Requests/sec`, `hours_back: 168`:

- 2,468 points returned
- **931 points (37.7%) had `delta_value == 0`**
- 0 points had a negative delta; 0 counter resets (`value` never decreased)
- Degrades sharply after the first day of the window:

| day | points | zero-delta |
|---|---|---|
| 08-06 | 683 | 2 |
| 08-07 | 493 | 122 |
| 08-08 | 225 | 192 |
| 08-09 | 319 | 123 |
| 08-10 | 230 | 126 |
| 08-11 | 217 | 156 |
| 08-12 | 224 | 149 |
| 08-13 | 77 | 61 |

A contiguous run where every `delta_value` was 0 while `value` advanced ~55,000 over ~70 minutes:

```
2026-08-11T13:06:47   20,056,930   delta 0
2026-08-11T13:12:33   20,060,956   delta 0
2026-08-11T13:19:10   20,066,328   delta 0
2026-08-11T13:28:31   20,072,557   delta 0
2026-08-11T13:36:14   20,077,796   delta 0
2026-08-11T13:45:12   20,083,510   delta 0
2026-08-11T13:51:40   20,087,811   delta 0
2026-08-11T13:59:17   20,092,702   delta 0
2026-08-11T14:08:34   20,100,946   delta 0
2026-08-11T14:17:02   20,111,780   delta 0
```

## Why this matters — it inverted a real result

I was testing whether a replica's CPU spikes were driven by request volume. Using `delta_value` as returned:

- "batches/sec p50 = 0.0" on four of seven days
- Pearson r between throughput and CPU = **−0.427** (negative)

Recomputing the rate myself from consecutive `value` readings, same data:

- batches/sec p50 = 10.3–12.2 every day, avg 11.1, p95 13.4
- Pearson r = **+0.304**

Same conclusion ("spikes are not volume-driven") but for the opposite reason. A consumer who trusted `delta_value` would have reported a negative correlation as a finding.

## Where it lives

The read is a faithful passthrough, so the defect is upstream of it — but the read also erases the distinction:

- `Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingTrendReader.cs` — `PerfmonTrendSql` selects `CAST(SUM(delta_cntr_value) AS bigint)` from `v_perfmon_stats` with no filtering.
- Same file, line 149: `reader.IsDBNull(2) ? 0 : reader.GetInt64(2)` — a **NULL delta is rendered as 0**, so "not computed" and "zero activity" become the same value at the API boundary. Line 148 does the same for `value`.

## The product already knows this data is untrustworthy

Two analysis consumers defend against exactly this and the MCP read does not:

- `Lite/Analysis/BaselineProvider.cs:294-300` — filters `delta_cntr_value >= 0` **and** discards a 0 whose predecessor was > 1000:
  ```sql
  AND   delta_cntr_value >= 0
  QUALIFY NOT (delta_cntr_value = 0
      AND COALESCE(LAG(delta_cntr_value) OVER (ORDER BY collection_time), 0) > 1000)
  ```
- `Lite/Analysis/AnomalyDetector.cs:712` — `AND delta_cntr_value >= 0`

So the spurious-zero behaviour is already known internally; `get_perfmon_trend` is the one path that exposes it unguarded to a caller who has no way to know.

## Suggested direction

Whichever of these is right is your call, but the current state is the one option that can't be defended:

1. Fix the writer so `delta_cntr_value` is either correct or **NULL** — never a fabricated 0.
2. Have the read distinguish them. If NULL must be projected, project it as NULL (or a `-1` sentinel, consistent with the house convention that an absent measurement must not read as a measured zero) rather than collapsing to 0 at `DarlingTrendReader.cs:149`.
3. At minimum, apply `BaselineProvider`'s guard in the MCP read so the three consumers agree on what the series means.

## Scope of what I checked

Read-side only, and on branch `feat/2216-incident-occurrence-counter` (whatever was checked out locally) — I did **not** trace the collector that writes `delta_cntr_value`, so I can't say whether it writes 0, writes NULL, or skips the row. I also did not check whether Lite's `get_perfmon_trend` has the same exposure, though `Lite/Services/LocalDataService.Perfmon.cs:99,149` uses the same unguarded `SUM(delta_cntr_value)`.

### PR DESCRIPTION
**Pull request by erikdarlingdata**
Fixes #2233. Fixes #2234.

## Measured first

I measured the live fleet store before changing the threshold, because #2233's premise ("trips on ordinary sweep jitter") is only actionable with a number. 99,717 consecutive perfmon gaps, 52 servers, 7 days:

| p50 | p90 | p99 | p99.9 | max |
|---|---|---|---|---|
| **299 s** | 580 s | 830 s | 1,190 s | 2,514 s |

The guard was **300 s**. It sat on the median. Share of *ordinary* gaps each candidate rejects:

| guard | rejects |
|---|---|
| **300 s (was)** | **50.0%** |
| 600 s | 8.3% |
| 900 s | 0.6% |
| 1,800 s | 0.0% |
| **3,600 s (now)** | **0.0%** |

So the guard fired during normal operation instead of after the restarts it was written for. And because a rejected gap returns `0`, and `0` is indistinguishable from a genuinely idle interval, it didn't merely lose data — **it invented quiet**. #2234 measured the same defect from the consumer end: 37.7% zero deltas while the cumulative counter advanced, which inverted a throughput/CPU correlation from **+0.304 to −0.427**.

One shared `DefaultMaxGapSeconds = 3600` replaces all 41 literals, so the policy can't drift collector by collector again. Five doc comments named `300` and would have become lies; `SeedLookback`'s doc actively *reasoned from* it ("a seed row from outside a ~5-minute window cannot produce a delta no matter what it cost to find") — that relationship inverts at an hour, and it now says so.

## Two more defects, found by checking the premise instead of assuming it

I had planned to expose `sample_interval_seconds` as the delta's denominator so a zero would be interpretable. Before shipping that I checked what the column actually holds:

```
zero_delta AND zero_interval    = 0
zero_delta AND nonzero_interval = 7,497
```

That refuted the design. The column isn't the delta's interval at all:

```csharp
.Value(60);   /* sample_interval_seconds — 1-minute collection interval */
```

**A hard-coded literal.** `PerfmonStatsCollector` asserted the configured cadence as fact on every row, against a measured median of 299 s. Every consumer divides by it — all of them via `NULLIF(sample_interval_seconds, 0)` — so each was computing a rate **up to 5× too high** from a denominator the collector had invented. Had I shipped my original plan, I'd have re-exported that lie under the friendlier name `covered_seconds`. `QueryStatsCollector` already measures its interval this way; perfmon was the outlier.

It measures now. That's what makes the read's half of #2234 possible: `delta_value = 0` is finally falsifiable from outside, since `interval 0` means unknown and `interval n` means idle.

**And a counter reset broke that invariant.** It returned delta `0` with a *real* interval — a claim that nothing happened for n seconds, in the one case where the claim is false. Reset now reports no interval either, so `interval == 0 ⟺ no delta was knowable`, exactly. That invariant is load-bearing for the whole (delta, interval) reading, and it was quietly untrue.

## Verification

Both test projects are `net10.0-windows` and need `Microsoft.WindowsDesktop.App`, so **the xunit pins cannot run on macOS** — they're CI-verified. To avoid shipping unverified logic I compiled the *shipped* `CollectorDeltaCalculator.cs` (linked, not copied, so it can't verify a stale variant) into a scratch net10.0 harness. 16/16, including the flip in both directions:

```
PASS  600s gap delta (was 0 pre-fix)      expected=55000 actual=55000
PASS  600s gap under OLD 300s policy      expected=0     actual=0
PASS  counter reset interval (was 120)    expected=0     actual=0
PASS  idle interval                       expected=180   actual=180
```

The last two are the invariant: reset and idle both carry a zero delta, and only the interval separates them. Builds clean (0 warnings) for both the Darling and Lite graphs; no existing test pinned the literal 60.

## Blast radius, stated plainly

- **Deltas across gaps of 300–3,600 s now report real accruals instead of 0.** They were always real; they were being suppressed. Anything with a threshold tuned against zero-suppressed deltas may see values it didn't before — that's the bug being removed, but it's a behavior change on a live fleet and worth your eye.
- **`sample_interval_seconds` on `perfmon_stats` changes meaning** from "always 60" to "measured, 0 when unknown". Existing consumers already `NULLIF(..., 0)`, so 0 becomes NULL rather than a divide-by-zero. Rows written before this keep their 60, so a rate over a window spanning the upgrade is only as good as its newest rows.
- **Not touched:** `Lite/Analysis/BaselineProvider.cs:299` carries a hand-rolled workaround for this exact bug — `QUALIFY NOT (delta_cntr_value = 0 AND LAG(delta_cntr_value) > 1000)`, filtering suspicious zeros. It should become unnecessary, but removing it is a separate behavioral change and belongs in its own PR. Other collectors' interval columns are unexamined; perfmon is the one #2234 named.

### CODE REVIEW

File: Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingTrendReader.cs [hunk 1]
```diff
@@ -48,7 +48,14 @@ public sealed record MemoryTrendPoint(
...
     /// both summed across the counter's instances at that collection (Lite's <c>PerfmonTrendPoint</c>).</summary>
-    public sealed record PerfmonTrendPoint(DateTime CollectionTime, long Value, long DeltaValue);
+    /// <summary>One perfmon point: the cumulative counter, the delta, and the wall-clock seconds that
+    /// delta actually covers, summed the same way. <c>SampleIntervalSeconds</c> is what makes a zero
+    /// readable: the collector reports 0 in exactly the cases where no delta was knowable (first
+    /// sighting, counter reset, gap past the policy), so (0, 0) is "unknown" while (0, n) is "genuinely
+    /// idle". Without it the two are the same number and a fabricated zero reads as quiet (#2234).
+    /// <para>Rows written before that fix carry a hard-coded 60 regardless of the real gap, so a rate
+    /// derived over a window spanning the upgrade is only as good as its newest rows.</para></summary>
+    public sealed record PerfmonTrendPoint(DateTime CollectionTime, long Value, long DeltaValue, long SampleIntervalSeconds);
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingTrendReader.cs]: Already fixed in `6df99060`, pushed before this review landed — `DocCommentHygieneTests.NoMemberCarriesTwoStackedSummaryBlocks` caught the same thing in CI. Merged into one summary that keeps the original's Lite parity note rather than deleting either block, since that test's own failure message warns most stacked summaries are displaced docs belonging to a *different* member.

I also replicated that rule locally (walking every `.cs` outside `bin`/`obj` for two `<summary>` openings in a single comment run), because both test projects are `net10.0-windows` and cannot run on this machine.

File: PerformanceMonitor.Collectors/PgWaitStatsCollector.cs [hunk 2]
```diff
@@ -179,10 +179,10 @@ is really an interval measurement. */
 
         var deltaWaits = context.Deltas.CalculateDelta(
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply PerformanceMonitor.Collectors/PgWaitStatsCollector.cs:180]: Fixed in `d4fe8928`. My sweep matched `300 s gap` and `300s gap`; this one spelled it `300-second gap policy` and survived — a decent illustration of why the number now lives in one constant instead of in prose. Re-swept for every spelling and this was the last.

File: Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingTrendReader.cs [hunk 3]
```diff
@@ -121,7 +128,8 @@ public static async Task<List<MemoryTrendPoint>> GetMemoryTrendAsync(
         SELECT
             collection_time,
             CAST(SUM(cntr_value) AS bigint) AS cntr_value,
-            CAST(SUM(delta_cntr_value) AS bigint) AS delta_cntr_value
+            CAST(SUM(delta_cntr_value) AS bigint) AS delta_cntr_value,
+            CAST(SUM(sample_interval_seconds) AS bigint) AS sample_interval_seconds
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingTrendReader.cs]: Confirmed and fixed in `d4fe8928` — you're right, and I verified it against the live fleet rather than reasoning about it. Distinct `instance_name` rows per `collection_time`, last 2 hours:

```
Log Bytes Flushed/sec      max=17  median=12
Log Flush Write Time (ms)  max=17  median=12
Log Flushes/sec            max=17  median=12
Transactions/sec           max=17  median=12
Lock Wait Time (ms)        max=15  median=15
Lock Waits/sec             max=15  median=15
```

So a caller following the doc comment I had *just written* would have derived rates **12-17x too low**. Same silent corruption this PR exists to remove, pointed the other way — and larger than the 5x error it was fixing. Changed to `MAX`, which also ignores a 0 from an instance seen for the first time while still reporting 0 when every instance is unknown.

You also identified why it slipped through: that pin was a string-contains check, so it passed on the wrong aggregate. It now names the aggregate, asserts `SUM(sample_interval_seconds)` is **absent**, and pins that `SUM(cntr_value)`/`SUM(delta_cntr_value)` stay sums so the fix can't be over-applied to the columns that genuinely are additive.

**Comment by erikdarlingdata** [comment]
Picked up the Viewer parity gap in `c48efa69` rather than filing a follow-up — you were right that this is the one place in the PR where "changed one but not its counterpart" genuinely applies, and the file's own doc comment claims a verbatim port from the Lite record I'd just changed.

`ViewerDataService.PerfmonTrendsSql` now projects `CAST(MAX(sample_interval_seconds) AS bigint)`, the record carries `SampleIntervalSeconds`, and the doc says the field rides along **unplotted** — the picker still plots the raw delta and derives no rate. The point is that whoever adds one finds a real denominator instead of reaching for a fixed cadence, which is the mistake that started this whole thread.

Worth noting the existing pin would not have caught a summed interval: it was called `PerfmonTrendsSql_SumsValueAndDelta_GroupedByCounterAndTime` and asserted "the two aggregates", so a third summed one would have sailed through. Renamed to `...ButTakesTheIntervalAsMax` and it now asserts `SUM(sample_interval_seconds)` is absent, matching the MCP-side guard.

That's all three surfaces (MCP read, Lite data layer, Viewer data layer) plus the shared collector, so the interval means the same thing everywhere it appears.

File: Lite/Services/LocalDataService.Perfmon.cs [hunk 4]
```diff
@@ -96,7 +96,8 @@ public async Task<List<PerfmonTrendPoint>> GetPerfmonTrendAsync(int serverId, st
 SELECT
     collection_time,
     SUM(cntr_value) AS cntr_value,
-    SUM(delta_cntr_value) AS delta_cntr_value
+    SUM(delta_cntr_value) AS delta_cntr_value,
+    MAX(sample_interval_seconds) AS sample_interval_seconds
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Lite/Services/LocalDataService.Perfmon.cs:100]: Added in `bfd4b9e6` — you're right that this was the one surface left unpinned, and the reason it mattered: Lite's queries are inline `CommandText`, not exposed constants, so there was nothing for a string pin to hold onto and nothing would have caught a reintroduced `SUM`.

New `Lite.Tests/PerfmonIntervalAggregationTests.cs` reads the source through `ParitySource`, which exists for exactly this cross-boundary case. Four facts:

- `MAX(sample_interval_seconds)` appears **twice** — a fix applied to only one of the single/batched pair is the likelier mistake
- `SUM(sample_interval_seconds)` appears nowhere
- `SUM(cntr_value)`/`SUM(delta_cntr_value)` each still appear twice, and `MAX(cntr_value)` nowhere — guarding the fix from being *over*-applied, which would report one instance's value as the whole counter
- Lite's MCP tool projects the field

Pre-flighted each assertion against the source by hand (2 / 0 / 2 / 2 / 0 / 1) since `Lite.Tests` is `net10.0-windows` and can't run on this machine.

File: PerformanceMonitor.Collectors/PerfmonStatsCollector.cs [hunk 5]
```diff
@@ -169,18 +169,27 @@ public override async ValueTask<List<Row>> ReadAsync(DbDataReader reader, Collec
 
     public override void WritePayload(Row row, ICollectorRowWriter writer, CollectorContext context)
     {
-        /* Delta for per-second counters — gap detection at 5min (5x the 1-min collection interval)
-           prevents inflated deltas after app restarts. Group/key/gap are the parity contract. */
+        /* Delta for per-second counters. Group/key/gap are the parity contract.
+
+           The interval is MEASURED, not assumed. This wrote a literal 60 — the configured one-minute
+           cadence — on every row, while the fleet's actual gap between perfmon sweeps runs a median of
+           299 s (p99 830 s, max 2,514 s over 99,717 samples). Every consumer that divides by it, and
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply PerformanceMonitor.Collectors/PerfmonStatsCollector.cs]: Corrected in `bfd4b9e6`. You're right, and I'd overstated it in the direction that flatters the change: the `NULLIF(sample_interval_seconds, 0)` idiom guards **query_stats'** interval, not perfmon's. Nothing in-product divides by perfmon's column — the MCP tools hand it to the caller, and the Viewer carries it unplotted per its own comment.

That also means the reassurance I'd written was hollow: there is no NULLIF standing between a caller and a 0 here, so the comment now says callers must treat 0 as unknown rather than dividing by it. The 5x claim is unchanged and still holds for anyone deriving a rate from the old literal 60.
